components/scream/src/physics/p3/p3_functions.hpp

```diff
@@ -585,16 +584,323 @@ struct Functions
+    const Int& nk_pack,
+    const uview_1d<const Spack>& oicldm,
+    const uview_1d<const Spack>& olcldm,
+    const uview_1d<const Spack>& orcldm,
+    const uview_1d<const Spack>& oexner,
+    const uview_1d<const Spack>& oth,
+    const uview_1d<Spack>& opratot,
+    const uview_1d<Spack>& oprctot,
+    const uview_1d<Spack>& prec,
+    const uview_1d<Spack>& mu_r,
+    const uview_1d<Spack>& odiag_ze,
+    const uview_1d<Spack>& ze_ice,
+    const uview_1d<Spack>& ze_rain,
+    const uview_1d<Spack>& odiag_effc,
+    const uview_1d<Spack>& odiag_effi,
+    const uview_1d<Spack>& odiag_vmi,
+    const uview_1d<Spack>& odiag_di,
+    const uview_1d<Spack>& odiag_rhoi,
+    const uview_1d<Spack>& ocmeiout,
+    const uview_1d<Spack>& oprain,
+    const uview_1d<Spack>& onevapr,
+    const uview_1d<Spack>& orflx,
+    const uview_1d<Spack>& osflx,
+    const uview_1d<Spack>& inv_icldm,
+    const uview_1d<Spack>& inv_lcldm,
+    const uview_1d<Spack>& inv_rcldm,
+    const uview_1d<Spack>& omu_c,
+    const uview_1d<Spack>& olamc,
+    const uview_1d<Spack>& inv_exner,
+    const uview_1d<Spack>& t,
+    const uview_1d<Spack>& oqv,
+    Scalar& prt_liq,
+    Scalar& prt_sol);
+
+  KOKKOS_FUNCTION
+  static void p3_main_pre_main_loop(
+    const MemberType& team,
+    const Int& nk_pack,
+    const bool& log_predictNc,
+    const Scalar& dt,
+    const uview_1d<const Spack>& opres,
+    const uview_1d<const Spack>& opdel,
+    const uview_1d<const Spack>& odzq,
+    const uview_1d<const Spack>& onpccn,
+    const uview_1d<const Spack>& oexner,
+    const uview_1d<const Spack>& inv_exner,
+    const uview_1d<const Spack>& inv_lcldm,
+    const uview_1d<const Spack>& inv_icldm,
+    const uview_1d<const Spack>& inv_rcldm,
+    const uview_1d<Spack>& t,
+    const uview_1d<Spack>& rho,
+    const uview_1d<Spack>& inv_rho,
+    const uview_1d<Spack>& qvs,
+    const uview_1d<Spack>& qvi,
+    const uview_1d<Spack>& sup,
+    const uview_1d<Spack>& supi,
+    const uview_1d<Spack>& rhofacr,
+    const uview_1d<Spack>& rhofaci,
+    const uview_1d<Spack>& acn,
+    const uview_1d<Spack>& oqv,
+    const uview_1d<Spack>& oth,
+    const uview_1d<Spack>& oqc,
+    const uview_1d<Spack>& onc,
+    const uview_1d<Spack>& oqr,
+    const uview_1d<Spack>& onr,
+    const uview_1d<Spack>& oqitot,
+    const uview_1d<Spack>& onitot,
+    const uview_1d<Spack>& oqirim,
+    const uview_1d<Spack>& obirim,
+    const uview_1d<Spack>& oxxlv,
+    const uview_1d<Spack>& oxxls,
+    const uview_1d<Spack>& oxlf,
+    const uview_1d<Spack>& qc_incld,
+    const uview_1d<Spack>& qr_incld,
+    const uview_1d<Spack>& qitot_incld,
+    const uview_1d<Spack>& qirim_incld,
+    const uview_1d<Spack>& nc_incld,
+    const uview_1d<Spack>& nr_incld,
+    const uview_1d<Spack>& nitot_incld,
+    const uview_1d<Spack>& birim_incld,
+    bool& log_nucleationPossible,
+    bool& log_hydrometeorsPresent);
+
+  KOKKOS_FUNCTION
+  static void p3_main_main_loop(
```

Comment:
I wonder if we shouldn't try to find a more descriptive name for this, like `p3_main_k_loop`, or `p3_main_microphys_loop` (and also avoid the use of `main` in the related `pre` and `post` functions). Not a huge deal, but if these functions are doing things that scientists refer to in given language, maybe we could appropriate that language.

Reply:
Yes, I'm hoping a scientist will chime-in with some name ideas, @pressel ?